lib/Sema/CSDiag.cpp

```diff
@@ -5241,14 +5243,31 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
       (isUnresolvedOrTypeVarType(fnExpr->getType()) ||
        (fnExpr->getType()->is<AnyFunctionType>() &&
         fnExpr->getType()->hasUnresolvedType()))) {
+    // FIXME: Prevent typeCheckChildIndependently from transforming expressions,
+    // because if we try to typecheck OSR expression with contextual type,
+    // it'll end up converting it into DeclRefExpr based on contextual info,
+    // instead let's try to get a type without applying and filter callee
+    // candidates later on.
     CalleeListener listener(CS->getContextualType());
-    fnExpr = typeCheckChildIndependently(callExpr->getFn(), Type(),
-                                         CTP_CalleeResult, TCC_ForceRecheck,
-                                         &listener);
-    if (!fnExpr) return true;
+
+    if (auto OSR = dyn_cast<OverloadSetRefExpr>(fnExpr)) {
+      ConcreteDeclRef decl = OSR->getReferencedDecl();
```

Comment:
One approach would be to assign nullptr explicitly to decl after asserting the OSR->getReferencedDecl() is nullptr. I'd rename decl to referencedDecl as well.

It doesn't look like the ASTVerifier is actually doing any checks on these Exprs, so it would be nice to add a check there as well.